### PR TITLE
Allow sparse color attachments and ignored FS outputs

### LINKED ISSUE #1250: Sparse colorAttachments in a render pass
Probably it should be possible to have "holes" in the colorAttachments array (e.g. have attachments 0 and 2, but not 1).

I know we discussed this a VERY long time ago, but right now in the API this is not expressible. We should make sure it's possible and change the API to support it (e.g. sequence of nullable, or put the attachment location inside the `GPURenderPassColorAttachmentDescriptor` like with `GPUBindGroupLayoutEntry`).

### LINKED ISSUE #2060: Feature request: ignore shader writes to color attachments
I would like to propose an ability to discard shader writes to certain color attachments if it is doable.

Use case from my codebase: 
I have a G-Buffer with Colors, Positions, Normals. I want to render opaque objects first and write associated g-buffer data. These data will be later used for (SS)AO (only opaque objects get AO computed). In the second pass I want to render transparent data into another color attachment BUT ignore writes of shader to positions and normals.

In this case I could use one shader with one pipeline with the desired functionality.

### PR DESCRIPTION
**Pull request by Kangz**
Fixes #1250
Fixes #2060

***
<a href="https://pr-preview.s3.amazonaws.com/Kangz/gpuweb/pull/2562.html" title="Last updated on Feb 9, 2022, 9:53 PM UTC (e822402)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2562/4cb81f5...Kangz:e822402.html" title="Last updated on Feb 9, 2022, 9:53 PM UTC (e822402)">Diff</a>

### CODE REVIEW

**Review by kvark — CHANGES_REQUESTED** [review]
blocking on investigation, which wasn't done yet

**Comment by toji** [comment]
No qualms about the API shape, assuming the group agrees this feature is feasible.

@kvark: Doing a quick review of the linked issues it's not clear to me what needs to be investigated still? https://github.com/gpuweb/gpuweb/issues/1250#issuecomment-1026905842 seems to confirm that it's technically feasible at least.

**Comment by Kangz** [comment]
Yeah I should have more clearly linked to that investigation. The behavior works on all APIs we targeted, even running with the debug layers.

**Review by kainino0x — COMMENTED** [review]
I just realized that using nullable here means both `null` and `undefined` work, which is great because I think it should mean _actual_ sparse arrays work (don't have to explicitly specify `null`). OTOH I may prefer @kvark's suggestion to add a `location` member to these dictionaries similar to vertex `attributeLocation`.

Also need to update:

- > For each |colorTarget| in |descriptor|.{{GPURenderPipelineDescriptor/fragment}}.{{GPUFragmentState/targets}}:

- > In alpha-to-coverage mode, an additional <dfn dfn>alpha-to-coverage mask</dfn>
of MSAA samples is generated based on the |alpha| component of the
fragment shader output value of the {{GPURenderPipelineDescriptor/fragment}}.{{GPUFragmentState/targets}}[0].

- Definition of `colorAttachments` should explain how it works

- `derive render targets layout from pass`

File: spec/index.bs [hunk 1]
```diff
@@ -5520,7 +5520,7 @@ dictionary GPUFragmentState : GPUProgrammableStage {
         Return `true` if all of the following requirements are met:
 
         - |descriptor|.{{GPUFragmentState/targets}}.length must be &le; 8.
-        - For each |colorState| layout descriptor in the list |descriptor|.{{GPUFragmentState/targets}}:
+        - For each non-`null` |colorState| layout descriptor in the list |descriptor|.{{GPUFragmentState/targets}}:
```
Comment (kvark) [inline spec/index.bs:5523]: Something needs to be said about the fact that the sets of targets need to match between the render pass and a pipeline. Here, the wording only happen to care about the non-null attachments, which is insufficient.

**Review by kvark — COMMENTED** [review]
Investigation approved:)
Considered alternative approach - not going to work out.
So we can proceed with this in principle, although @kainino0x had ideas on making the shape of the API more usable.
One last thing to address

**Comment by kainino0x** [comment]
> So we can proceed with this in principle, although @kainino0x had ideas on making the shape of the API more usable.

If we want to allow a `{0: x, 2: y}` syntax we could use a `record<>` but I don't think it's important. Like I said above it introduces "optional" syntax and that's ugly. Maybe we could add it later as a nonbreaking change if we really wanted to.

**Review by kainino0x — APPROVED** [review]
no further comments, happy to go with this design option

**Comment by Kangz** [comment]
Sorry, I'm not going to be able to address review comments anytime soon. Can someone take over when/if the group decides to go forward with this?

**Comment by kvark** [comment]
@Kangz I'll push to this PR and merge

**Review by kvark — APPROVED** [review]

File: spec/index.bs (hunk 1 above)
Comment (kvark) [reply spec/index.bs:5523]: Addressed myself in another commit

**Review by kainino0x — APPROVED** [review]

**Review by toji — APPROVED** [review]

**Comment by kainino0x** [comment]
Ah, merging this closed 2 issues. Are they both fully resolved? #2060 in particular.

**Comment by Kangz** [comment]
Yeah #2060 is fixed: there's the note, and the [fragment processing description](https://gpuweb.github.io/gpuweb/#fragment-processing) says:

> Set fragment.[colors](https://gpuweb.github.io/gpuweb/#fragment-colors) to the user-specified [pipeline output](https://gpuweb.github.io/gpuweb/wgsl/#pipeline-output) values from the shader.

Which is imprecise bit supports sparse attachments correctly. (ideally it should be a loop that says, for each attachment color, take the value if it is present, or fill with (0, 0, 0, 1)).

**Pull request merged**